wallet, refactor: return out-params of CreateTransaction() as optional struct

theStack opened the pull request:
The method `CWallet::CreateTransaction` currently returns several values in the form of out-parameters:
* the actual newly created transaction (`CTransactionRef& tx`)
* its required fee (`CAmount& nFeeRate`)
* the position of the change output (`int& nChangePosInOut`) -- as the name suggests, this is both an in- and out-param

By returning these values in an optional structure (which returns no value a.k.a. `std::nullopt` if an error occured), the interfaces is shorter, cleaner (requested change position is now in-param and can be passed by value) and callers don't have to create dummy variables for results that they are not interested in. 

Note that the names of the replaced out-variables were kept in `CreateTransactionInternal` to keep the diff minimal. Also, the fee calculation data (`FeeCalculation& fee_calc_out`) would be another candidate to put into the structure, but `FeeCalculation` is currently an opaque data type in the wallet interface and I think it should stay that way.

As a potential follow-up, I think it would make sense to also do the same refactoring for `CWallet::FundTransaction`, which has a very similar parameter structure.

Suggested by laanwj in https://github.com/bitcoin/bitcoin/pull/20588#issuecomment-739838428.

theStack committed:
wallet: CreateTransactionInternal(): return out-params as (optional) struct

theStack committed:
wallet: CreateTransaction(): return out-params as (optional) struct

pox commented on src/wallet/feebumper.cpp:
```diff
@@ -215,21 +215,19 @@ Result CreateRateBumpTransaction(CWallet& wallet, const uint256& txid, const CCo
     bilingual_str fail_reason;
     FeeCalculation fee_calc_out;
-    if (!wallet.CreateTransaction(recipients, tx_new, fee_ret, change_pos_in_out, fail_reason, new_coin_control, fee_calc_out, false)) {
+    auto txr = wallet.CreateTransaction(recipients, /* No requested location for change */ -1, fail_reason, new_coin_control, fee_calc_out, false);
```
While it's arguably not, strictly speaking, part of the proposed change, the comment about the change position parameter that's being maintained here is a bit misleading. This is a bit nit-picky, so bare with me please, but I think it's more appropriate to:
1. Use a const instead of inline comment in order to convey the meaning. e.g. `const int RANDOM_CHANGE_POSITION = -1`). This value is used more than once, so another reason to use a const.
2. Convey that passing `-1` means "actively set the position to a random one" rather than "no requested location" which doesn't really let the reader know what's going to happen.

theStack replied:
Thanks a lot for reviewing, good point! I probably was a bit over-enthusiastic in getting rid of as many variables as possible before the call to `CreateTransaction` ;-) Introducing a `RANDOM_CHANGE_POSITION` constant makes a lot of sense to me, see latest force-push of the second commit. Not sure yet if putting the constant into the header is a good idea, as it is only used for one single method and hence many modules that don't need the constant would see it as well. Probably a candidate for a follow-up PR?

pox commented on src/wallet/wallet.cpp:
```diff
@@ -3099,42 +3103,34 @@ bool CWallet::CreateTransactionInternal(
-                nFeeRet = nFeeRet2;
-                nChangePosInOut = nChangePosInOut2;
-            }
+            const bool use_aps = txr_grouped->fee <= txr_ungrouped->fee + m_max_aps_fee;
```
This is a lot more readable indeed! 👍

pox reviewed (commented):
ACK not tested.

theStack commented:
Force-pushed with the proposed changes by @pox (see https://github.com/bitcoin/bitcoin/pull/20640#discussion_r546194751), introducing new constants `RANDOM_CHANGE_POSITION` to be used as second parameter for `CWallet::CreateTransaction`.

practicalswift commented on src/wallet/feebumper.cpp:
```diff
@@ -215,21 +215,20 @@ Result CreateRateBumpTransaction(CWallet& wallet, const uint256& txid, const CCo
     bilingual_str fail_reason;
     FeeCalculation fee_calc_out;
-    if (!wallet.CreateTransaction(recipients, tx_new, fee_ret, change_pos_in_out, fail_reason, new_coin_control, fee_calc_out, false)) {
+    auto txr = wallet.CreateTransaction(recipients, RANDOM_CHANGE_POSITION, fail_reason, new_coin_control, fee_calc_out, false);
```
Nit: Using `auto` here obscures the fact that `txr` is an optional. I think being explicit about the optionality is better in order to avoid mistakes in the future.

Applies throughout this PR.

theStack replied:
Thanks for reviewing! Agree that being explicit is better here. Done.

stackman27 commented:
Code review and built successful. Also, definitely a good idea to put `FeeCalculation& fee_calc_out` into this structure

stackman27 reviewed (approved):
ACK Code review and built successful. Also, definitely a good idea to put `FeeCalculation& fee_calc_out ` into this structure

pox commented:
reACK 5865c38
🙏

theStack commented:
Force-pushed with changes [as suggested by practicalswift](https://github.com/bitcoin/bitcoin/pull/20640#discussion_r549173268), using explicit datatype `std::optionial<CreatedTransactionalResult>` rather instead of `auto` for storing return values of `CreateTransaction{Internal}()`. Should be straight-forward to re-review, see [range-diff](https://github.com/bitcoin/bitcoin/compare/5865c38c6a3dd6a98ff3d03ad5e49c550db49a21..c32e2b1ccd1145082c29d80e62affa66e246e2fc).

theStack commented:
Rebased on master.

theStack commented:
Rebased on master and changed code to return uninitialized values explicitely via `return std::nullopt` rather than `return {}` (see https://github.com/bitcoin/bitcoin/pull/21498, https://github.com/bitcoin/bitcoin/pull/21415#issuecomment-800236640 for further explanation why this is preferred).

theStack commented:
Rebased on master.

theStack commented:
Rebased on master again. (This time it was a bit more complex, as the involved methods have moved from `wallet/wallet.cpp` to `wallet/spend.cpp`).

theStack commented:
Rebased on master.

fanquake commented:
@achow101 @glozow any Concept ACK / NACK thoughts on this PR?

achow101 commented:
Concept ACK

theStack commented:
Rebased on master.

laanwj commented:
Concept ACK. Returning things is much cleaner for out-only arguments than using mutable arguments (which would ideally be reserved for inout-arguments only).

furszy reviewed (commented):
First look, concept ACK.

Wasn't aware of this PR and touched the same piece of code in #24845.

-> PR where I introduced two generic versions of the optional return object called `OperationResult` and `CallResult<T>` (basically classes that encapsulate the function result: the result object, or in case of failure, the failure reason), which later can be used everywhere in the sources to cleanup similar boilerplate code where we pass a lot of ref args with a single return optional struct like the one introduced here.

So, combining both PRs, we will just return a `CallResult<CreatedTransactionResult>` which will cleanup the code further.

Will do a deeper code review soon.

achow101 commented on src/wallet/spend.h:
```diff
@@ -95,9 +97,9 @@ struct CreatedTransactionResult
+ * @note passing nChangePos as -1 will result in setting a random position
  */
-bool CreateTransaction(CWallet& wallet, const std::vector<CRecipient>& vecSend, CTransactionRef& tx, CAmount& nFeeRet, int& nChangePosInOut, bilingual_str& error, const CCoinControl& coin_control, FeeCalculation& fee_calc_out, bool sign = true);
+std::optional<CreatedTransactionResult> CreateTransaction(CWallet& wallet, const std::vector<CRecipient>& vecSend, int nChangePos, bilingual_str& error, const CCoinControl& coin_control, FeeCalculation& fee_calc_out, bool sign = true);
```
In 5238022161a1ecd361916c6f425439e7c7604758 "wallet: CreateTransaction(): return out-params as (optional) struct"

nit: new variables should be snake_case.

Perhaps we should consider changing this to `std::optional<int>` and get rid of using `-1` as a magic number.

achow101 reviewed (commented):
ACK 5238022161a1ecd361916c6f425439e7c7604758

theStack commented:
Rebased on master. Thanks to @furszy for getting the ball rolling on this PR again by pinging me up on IRC (I was planning to wait for #25003, but it seems that this will not be merged soon, as it was suggested to remove the affected tracepoint completely).

murchandamus reviewed (commented):
ACK 4c5ceb040cf50d24201903a9200fb23be88d96fb

achow101 commented:
re-ACK 4c5ceb040cf50d24201903a9200fb23be88d96fb

w0xlt reviewed (approved):
crACK https://github.com/bitcoin/bitcoin/pull/20640/commits/4c5ceb040cf50d24201903a9200fb23be88d96fb
